pkg/controller/controller_manager.go

```diff
@@ -75,55 +75,154 @@ func (cm *ControllerManager) GetReconciler(name string) Reconciler {
-	reconciler, queue, err := reconcilerGetter()
+func (cm *Manager) AddReconcilerOrDie(name string, reconcilerGetter func() (Reconciler, error)) {
+	reconciler, err := reconcilerGetter()
 	if err != nil {
-		log.Logger.Errorw("unable to create reconciler", "name", name, "error", err)
+		klog.ErrorS(err, "unable to create reconciler", "name", name)
 		os.Exit(1)
 	}
 
 	cm.AddController(&Controller{
-		Name:  name,
-		Do:    reconciler,
-		Queue: queue,
+		Name: name,
+		Do:   reconciler,
 	})
 }
 
-func (cm *ControllerManager) Start(ctx context.Context) error {
+func (cm *Manager) Start(ctx context.Context) error {
+	cm.Lock()
+	defer cm.Unlock()
+
 	if cm.started {
-		return errors.New("controller manager was started more than once")
+		return errors.New("console controller manager was started more than once")
 	}
 
-	for _, ctrl := range cm.Controllers {
-		controller := ctrl
-		jitterValue := time.Duration(rand.Int63n(int64(cm.Jitter)))
-		cm.Socket.AddPublisher(controller.Do.GetPublisher())
+	go cm.startSupervised(ctx)
 
-		go func() {
-			defer controller.Do.ShutdownQueue()
-			defer controller.Do.WipeCache()
+	_ = helpers.BackgroundPollUntilContextCancel(ctx, cm.PollInterval, true, false, func(_ context.Context) (bool, error) {
+		if err := cm.Socket.Join(); err != nil {
+			klog.ErrorS(err, "unable to connect")
+		}
 
-			pollInterval := cm.Refresh
-			if controllerPollInterval := controller.Do.GetPollInterval(); controllerPollInterval > 0 {
-				pollInterval = controllerPollInterval
-			}
-			pollInterval += jitterValue
-			_ = wait.PollUntilContextCancel(context.Background(), pollInterval, true, func(_ context.Context) (done bool, err error) {
-				return controller.Do.Poll(ctx)
-			})
-		}()
+		// never stop
+		return false, nil
+	})
+
+	cm.started = true
+	return nil
+}
+
+func (cm *Manager) startSupervised(ctx context.Context) {
+	wg := &sync.WaitGroup{}
+	wg.Add(len(cm.Controllers))
 
+	for _, ctrl := range cm.Controllers {
 		go func() {
-			controller.Start(ctx)
+			defer wg.Done()
+			cm.startControllerSupervised(ctx, ctrl)
 		}()
 	}
 
+	<-ctx.Done()
+	klog.InfoS("Shutdown signal received, waiting for all controllers to finish", "name", "console-manager")
+	wg.Wait()
+	klog.InfoS("All controllers finished", "name", "console-manager")
+}
+
+func (cm *Manager) startControllerSupervised(ctx context.Context, ctrl *Controller) {
+	internalCtx, cancel := context.WithCancel(ctx)
+	wg := &sync.WaitGroup{}
+
+	// Recheck the controller heartbeat every 30 seconds.
+	heartbeatCheckInterval := 30 * time.Second
+	// Make last heartbeat deadline 3 times the time of regular poll.
+	// It means that the controller poller skipped at least the last 2 scheduled polls.
+	// It could indicate that the controller poll might have died and controller should be restarted.
+	lastHeartbeatDeadline := 3 * (cm.PollInterval + cm.Jitter)
```

Comment:
I feel like there should be an option to change that. Perhaps we should add argument for it and check if it is not too short during initialisation, or have an argument for number of scheduled polls. I am also wondering if some issues would not appear if someone would use short interval and jitter as API requests can take a while.

Reply:
It's not very realistic but still I'd add option to change that later.